File: jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/accmod/CheckAccess.java
package org.jenkinsci.gradle.plugins.jpi2.accmod;

import org.gradle.workers.WorkAction;
import org.kohsuke.accmod.impl.Checker;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.List;
import java.util.Properties;

/**
 * Work action that runs {@code kohsuke.accmod.Checker} against a compiled class directory
 * to enforce {@code @Restricted} access-modifier rules.
 */
public abstract class CheckAccess implements WorkAction<CheckAccessParameters> {
    private static final Logger LOGGER = LoggerFactory.getLogger(CheckAccess.class);

    /** Required by Gradle's worker injection infrastructure. */
    @Inject
    public CheckAccess() {
    }

    @Override
    public void execute() {
        List<URL> toScan = getParameters().getClasspathToScan().getFiles().stream()
                .map(file -> {
                    try {
                        return file.toURI().toURL();
                    } catch (Exception e) {
                        throw new IllegalStateException("Could not convert file to URL: " + file, e);
                    }
                })
                .toList();
        URL[] urls = toScan.toArray(new URL[0]);

        URLClassLoader loader = new URLClassLoader(urls, getClass().getClassLoader());
        InternalErrorListener listener = new InternalErrorListener();
        Properties props = new Properties();
        getParameters().getPropertiesForAccessModifier().get().forEach(props::put);

        try {
            Checker checker = new Checker(loader, listener, props, new MavenLoggingBridge());
            checker.check(getParameters().getDirToCheck().get().getAsFile());
        } catch (IOException e) {
            throw new IllegalStateException("Failed to run restricted API checks", e);
        }

        try {
            var outputFile = getParameters().getOutputFile().get().getAsFile().toPath();
            Files.createDirectories(outputFile.getParent());
            Files.writeString(outputFile, listener.errorMessage(), StandardCharsets.UTF_8);
        } catch (Exception e) {
            throw new IllegalStateException("Failed to write access modifier report", e);
        }

        if (listener.hasErrors()) {
            if (getParameters().getIgnoreFailures().get()) {
                LOGGER.warn(listener.errorMessage());
            } else {
                LOGGER.error(listener.errorMessage());
                throw new RestrictedApiException();
            }
        }
    }
}

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccess.kt
package org.jenkinsci.gradle.plugins.accmod

import org.gradle.workers.WorkAction
import org.kohsuke.accmod.impl.Checker
import org.slf4j.LoggerFactory
import java.net.URLClassLoader
import java.util.*

abstract class CheckAccess : WorkAction<CheckAccessParameters> {
    private companion object {
        private val LOGGER = LoggerFactory.getLogger(CheckAccess::class.java)
    }

    override fun execute() {
        val toScan = parameters.classpathToScan.files.map { it.toURI().toURL() }
        val array = toScan.toTypedArray()
        val loader = URLClassLoader(array, javaClass.classLoader)
        val listener = InternalErrorListener()
        val props = Properties()
        for ((key, value) in parameters.propertiesForAccessModifier.get()) {
            props[key] = value
        }
        val checker = Checker(loader, listener, props, MavenLoggingBridge)
        checker.check(parameters.dirToCheck.asFile.get())
        parameters.outputFile.asFile.get().writeText(listener.errorMessage())
        if (listener.hasErrors()) {
            if (parameters.ignoreFailures.get()) {
                LOGGER.warn(listener.errorMessage())
            } else {
                LOGGER.error(listener.errorMessage())
                throw RestrictedApiException()
            }
        }
    }
}
